### PR TITLE
NMS-12586: Fix restoring of facts on DroolsCorrelationEngine reload

### PR DESCRIPTION
**Pull request by cgorantla**
Marshal objects that are in the session before reload.
Unmarshal with specific class loader while restoring

* JIRA (Issue Tracker): http://issues.opennms.org/browse/NMS-12586
* Bamboo (Continuous Integration): https://bamboo.opennms.org/

### CODE REVIEW

File: opennms-correlation/drools-correlation-engine/src/main/java/org/opennms/netmgt/correlation/drools/DroolsCorrelationEngine.java [hunk 1]
```diff
@@ -505,17 +505,50 @@ void saveFacts( ) {
...
+        }
+        return null;
+    }
+
+    private void marshalAndSaveFact(FactHandle factHandle) {
+        Object factObject = m_kieSession.getObject(factHandle);
+        try {
+            factObjects.put(DroolsStreamUtils.streamOut(factObject), factObject.getClass());
+        } catch (IOException e) {
+            LOG.error("Exception while marshalling fact {}", e);
```
Comment (j-white) [inline opennms-correlation/drools-correlation-engine/src/main/java/org/opennms/netmgt/correlation/drools/DroolsCorrelationEngine.java]: Add factObject & factObject class to exception.

File: opennms-correlation/drools-correlation-engine/src/main/java/org/opennms/netmgt/correlation/drools/DroolsCorrelationEngine.java [hunk 2]
```diff
@@ -505,17 +505,50 @@ void saveFacts( ) {
...
+            String packageName = clazz.getPackage().getName();
+            String className = clazz.getSimpleName();
+            FactType factType = getFactType(packageName, className);
+            if (factType != null) {
+                m_kieSession.insert(DroolsStreamUtils.streamIn(factBytes, factType.getFactClass().getClassLoader()));
+            } else {
+                m_kieSession.insert(DroolsStreamUtils.streamIn(factBytes, clazz.getClassLoader()));
+            }
+        } catch (IOException | ClassNotFoundException e) {
+            LOG.error("Exception while unmarshalling fact {} with class {}", factBytes, clazz.getCanonicalName());
```
Comment (j-white) [inline opennms-correlation/drools-correlation-engine/src/main/java/org/opennms/netmgt/correlation/drools/DroolsCorrelationEngine.java]: Remove factBytes or replace with factBytes.length from exception.

File: opennms-correlation/drools-correlation-engine/src/test/opennms-home/etc/drools-engine.d/persistState/PersistState.drl [hunk 3]
```diff
@@ -4,13 +4,22 @@ import org.opennms.netmgt.correlation.drools.DroolsCorrelationEngine;
...
 
 declare Thing
 	name: String
 end
 
+declare PersistedTestFact
+  factId: String
+end
+
+query "find PersistedTestFact by factId" (String searchFactId)
```
Comment (j-white) [inline opennms-correlation/drools-correlation-engine/src/test/opennms-home/etc/drools-engine.d/persistState/PersistState.drl]: query, nice

Comment (cgorantla) [reply opennms-correlation/drools-correlation-engine/src/test/opennms-home/etc/drools-engine.d/persistState/PersistState.drl]: This is all taken from @UberPinguin 's PR :-)

**Pull request merged**